types/wlr_drm_lease_v1.c

```diff
@@ -0,0 +1,704 @@
+	wl_list_for_each(connector, &device->connectors, link) {
+		drm_lease_connector_v1_send_to_client(connector, device_resource);
+	}
+
+	wp_drm_lease_device_v1_send_done(device_resource);
+}
+
+static void handle_output_destroy(struct wl_listener *listener, void *data) {
+	struct wlr_drm_lease_connector_v1 *conn = wl_container_of(listener, conn,
+			destroy);
+	wlr_log(WLR_DEBUG, "Handle destruction of output %s", conn->output->name);
+	wlr_drm_lease_v1_manager_withdraw_output(conn->device->manager, conn->output);
+}
+
+bool wlr_drm_lease_v1_manager_offer_output(
+		struct wlr_drm_lease_v1_manager *manager, struct wlr_output *output) {
+	assert(manager && output);
+	assert(wlr_output_is_drm(output));
+
+	wlr_log(WLR_DEBUG, "Offering output %s", output->name);
+
+	struct wlr_drm_lease_device_v1 *device = NULL, *tmp_device;
+	wl_list_for_each(tmp_device, &manager->devices, link) {
+		if (tmp_device->backend == output->backend) {
+			device = tmp_device;
+			break;
+		}
+	}
+
+	if (!device) {
+		wlr_log(WLR_ERROR, "No wlr_drm_lease_device_v1 associated with the "
+				"offered output");
+		return false;
+	}
+
+	struct wlr_drm_lease_connector_v1 *tmp_connector;
+	wl_list_for_each(tmp_connector, &device->connectors, link) {
+		if (tmp_connector->output == output) {
+			wlr_log(WLR_ERROR, "Output %s has already been offered",
+					output->name);
+			return false;
+		}
+	}
+
+	struct wlr_drm_lease_connector_v1 *connector =
+			calloc(1, sizeof(struct wlr_drm_lease_connector_v1));
+	if (!connector) {
+		wlr_log(WLR_ERROR, "Failed to allocate wlr_drm_lease_connector_v1");
+		return false;
+	}
+
+	connector->output = output;
+	connector->device = device;
+
+	connector->destroy.notify = handle_output_destroy;
+	wl_signal_add(&output->events.destroy, &connector->destroy);
+
+	wl_list_init(&connector->resources);
+	wl_list_insert(&device->connectors, &connector->link);
+
+	struct wl_resource *resource;
+	wl_resource_for_each(resource, &device->resources) {
+		drm_lease_connector_v1_send_to_client(connector, resource);
+	}
+
+	return true;
+}
+
+void wlr_drm_lease_v1_manager_withdraw_output(
+	struct wlr_drm_lease_v1_manager *manager, struct wlr_output *output) {
+	assert(manager && output);
+
+	wlr_log(WLR_DEBUG, "Withdrawing output %s", output->name);
+
+	struct wlr_drm_lease_device_v1 *device = NULL, *tmp_device;
+	wl_list_for_each(tmp_device, &manager->devices, link) {
+		if (tmp_device->backend == output->backend) {
+			device = tmp_device;
+			break;
+		}
+	}
+
+	if (!device) {
+		wlr_log(WLR_ERROR, "No wlr_drm_lease_device_v1 associated with the "
+				"given output");
+		return;
+	}
+
+	struct wlr_drm_lease_connector_v1 *connector = NULL, *tmp_conn;
+	wl_list_for_each(tmp_conn, &device->connectors, link) {
+		if (tmp_conn->output == output) {
+			connector = tmp_conn;
+			break;
+		}
+	}
+
+	if (!connector) {
+		wlr_log(WLR_DEBUG, "Output has not been offered");
+		return;
+	}
+
+	if (connector->active_lease) {
```

Comment:
Maybe we should move this block to `drm_lease_connector_v1_destroy`, or at least add an `assert(connector->active_lease == NULL)`. Destroying the connector without the lease would leave a dangling connector pointer behind.

Reply:
Moved it to `drm_lease_connector_v1_destroy`